pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/dictionary/v2/BaseMutableDictionary.java

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.realtime.impl.dictionary.v2;
+
+import com.linkedin.pinot.core.segment.index.readers.Dictionary;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nonnull;
+
+
+/**
+ * The class <code>BaseMutableDictionary</code> is the implementation of the mutable dictionary required by REALTIME
+ * consuming segments.
+ * <p>The implementation needs to be thread safe for single writer multiple readers scenario.
+ * <p>We can assume the readers always first get the dictionary id for a value, then use the dictionary id to fetch the
+ * value later, but not reversely. So whenever we return a valid dictionary id for a value, we need to ensure the value
+ * can be fetched by the dictionary id returned.
+ */
+public abstract class BaseMutableDictionary implements Dictionary {
+  private static final int MAX_CARDINALITY = 100_000_000;
+  private static final int INITIAL_DICTIONARY_SIZE = 10_000;
```

Comment:
Also, can we use fastutils here? It appears that it works with one writer and multiple readers, but please verify